packages/artifact/src/upload-http-client.ts

```diff
@@ -0,0 +1,337 @@
+  await Promise.all(
+    parallelUploads.map(async () => {
+      while (uploadedFiles < filesToUpload.length) {
+        const currentFileParameters = parameters[uploadedFiles]
+        uploadedFiles += 1
+        if (abortPendingFileUploads) {
+          failedItemsToReport.push(currentFileParameters.file)
+          continue
+        }
+
+        const uploadFileResult = await uploadFileAsync(currentFileParameters)
+        fileSizes += uploadFileResult.successfulUploadSize
+        if (uploadFileResult.isSuccess === false) {
+          failedItemsToReport.push(currentFileParameters.file)
+          if (!continueOnError) {
+            // Existing uploads will be able to finish however all pending uploads will fail fast
+            abortPendingFileUploads = true
+          }
+        }
+      }
+    })
+  )
+
+  // eslint-disable-next-line no-console
+  console.log(`Total size of all the files uploaded is ${fileSizes} bytes`)
+  return {
+    size: fileSizes,
+    failedItems: failedItemsToReport
+  }
+}
+
+/**
+ * Asynchronously uploads a file. If the file is bigger than the max chunk size it will be uploaded via multiple calls
+ * @param {UploadFileParameters} parameters Information about the file that needs to be uploaded
+ * @returns The size of the file that was uploaded in bytes along with any failed uploads
+ */
+async function uploadFileAsync(
+  parameters: UploadFileParameters
+): Promise<UploadFileResult> {
+  const fileSize: number = fs.statSync(parameters.file).size
+  const parallelUploads = [...new Array(parameters.concurrency).keys()]
+  let offset = 0
+  let isUploadSuccessful = true
+  let failedChunkSizes = 0
+  let abortFileUpload = false
+
+  await Promise.all(
+    parallelUploads.map(async () => {
+      while (offset < fileSize) {
+        const chunkSize = Math.min(fileSize - offset, parameters.maxChunkSize)
+        if (abortFileUpload) {
+          // if we don't want to continue on error, any pending upload chunk will be marked as failed
+          failedChunkSizes += chunkSize
+          continue
+        }
+
+        const start = offset
+        const end = offset + chunkSize - 1
+        offset += parameters.maxChunkSize
+        const chunk: NodeJS.ReadableStream = fs.createReadStream(
+          parameters.file,
+          {
+            start,
+            end,
+            autoClose: false
+          }
+        )
+
+        const result = await uploadChunk(
+          parameters.restClient,
+          parameters.resourceUrl,
+          chunk,
+          start,
+          end,
+          fileSize
+        )
+
+        if (!result) {
+          /**
+           * Chunk failed to upload, report as failed but continue if desired. It is possible that part of a chunk was
+           * successfully uploaded so the server may report a different size for what was uploaded
+           **/
+          isUploadSuccessful = false
+          failedChunkSizes += chunkSize
+          if (!parameters.continueOnError) {
```

Comment:
We should not allow this.  If a chunk failed we should either retry or fail the file - `continueOnError` means if a file failed to upload, we still allow uploading other files; but the atomic unit should be file, a file will either upload successfully or not - we should not allow partially succeeded single file.

Reply:
Good point. I'll go ahead and change this